Port to Funogram

ForNeVeR opened the pull request.

ForNeVeR committed:
Dependencies: switch to Funogram

ForNeVeR committed:
Port to Funogram

Szer commented on src/VahterBanBot/Bot.fs:
```diff
@@ -697,14 +715,14 @@ let vahterMarkedAsSpam
     %onCallbackActivity.SetTag("callbackId", callbackQuery.Data)
     
-    let callbackId = Guid.Parse callbackQuery.Data
+    let callbackId = Guid.Parse(callbackQuery.Data |> Option.defaultValue "")
```
я бы предложил с грохотом упать если у нас не callbackQuery.Data.
Оно конечно и так упадёт, ноу же с ошибкой парсинга гуида, что чуть менее понятно

Szer commented on src/VahterBanBot/Converters.fs:
```diff
@@ -0,0 +1,252 @@
+module internal Converters =
+  open Funogram.StringUtils
+
+  let private unixEpoch = DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc)
```
да неужто встроенного нет!

Szer commented on src/VahterBanBot/Program.fs:
```diff
@@ -91,17 +88,17 @@ let builder = WebApplication.CreateBuilder()
     .AddSingleton<MachineLearning>()
     .AddHostedService<MachineLearning>(fun sp -> sp.GetRequiredService<MachineLearning>())
+    // TODO[F]: Figure out these
     .AddHttpClient("telegram_bot_client")
```
а я не знаю зачем это, делал по примеру из Telegram.Bot

Может и нахер не надо

Szer commented on src/VahterBanBot/Program.fs:
```diff
@@ -193,22 +190,23 @@ app.UseGiraffe(webApp)
 let server = app.RunAsync()
 
 // Dev mode only
```
я б удалил в целом, это Кирилл для дебага делал, но я хз.

В целом есть планы перейти на полноценный поллинг, но он не так будет выглядеть

Szer commented on src/VahterBanBot/Utils.fs:
```diff
@@ -28,12 +32,13 @@ let getEnvOrWith name defaultValue action =
+    match s with
+    | None -> null
+    | Some s when s.StartsWith "@" -> s
+    | Some s -> s
```
багуля?
```suggestion
    | Some s -> "@" + s
```

Szer reviewed (commented):
Очень хорошо! Мне нравится

Dolfik1 commented on src/VahterBanBot/Utils.fs:
```diff
@@ -57,29 +62,48 @@ type Task<'x> with
+        |> Option.orElse msg.Caption
+        |> Option.defaultValue ""
+        
+    member msg.MessageId32 = // the documentation guarantees that this always fits into int32, Funogram disagrees
```
У них в документации Type всегда Integer. Для некоторых типов в документации указано:

> Unique identifier for this user or bot. This number may have more than 32 significant bits and some programming languages may have difficulty/silent defects in interpreting it. But it has at most 52 significant bits, so a 64-bit integer or double-precision float type are safe for storing this identifier.

Я решил, что безопаснее везде использовать int64, чем надеяться на то, что генератор корректно определит тип на основе desciption, и вообще, что в документации это будет указано.

ForNeVeR committed:
Program: get rid of UsePolling parameter

ForNeVeR committed:
Tests: fix compilation after Funogram migration

ForNeVeR committed:
Fix some tests after migration

ForNeVeR committed:
More robust testing (#62)

Tests: more robust logging for failed containers, destroy test container before startup

ForNeVeR committed:
Docker: update to .NET SDK 8.0.402

ForNeVeR committed:
Funogram: enable test API

ForNeVeR commented:
No longer needed after #188.
